Support block-readable entries

ekinanp opened the pull request:
External plugins will be handled in a follow-up PR. We should also file a separate issue for the volume package; I didn't implement BlockReadable for the volume package because it isn't as easy to do as e.g. S3 objects or GCP storage objects.

Supports https://github.com/puppetlabs/wash/issues/619

ekinanp committed:
Refactor the Read action to allow for block readable entries

At a high level, a "read" invocation should return the entry’s content. However,
some entries’ content can be read in blocks (e.g. S3/GCP storage objects),
paving the way for more optimized reads. We categorize the latter group of
entries as "block readable entries". This implies two different signatures for the
"read" action:

* A "default" signature that returns all of the entry’s content

* A "block readable" signature that returns the requested block of the entry’s
content (for block readable entries)

A readable entry can only implement one or the other; it cannot implement both
signatures.

To support these changes, this commit

* Changes the "Readable" interface’s signature to "Read(ctx) ([]byte, error)". This
represents the "default" signature.

* Adds a new "BlockReadable" interface with signature "Read(ctx, size, offset) ([]byte, error)".
This represents the "block readable" signature.

* Updates the plugin.Action type to support method signatures

* Changes "plugin.Open(ctx, e)" to "plugin.Read(ctx, e, size, offset)". The latter, general signature
is what the Wash server needs when reading the entry’s content.

* Updates "plugin.cachedOpen(ctx, e)" to "plugin.cachedRead(ctx, e)". The latter signature
centralizes the high-level "returns the entry’s content" nature of "read" while also making
It easier to (in the future) support cached block reads. This is achieved via the "entryContent"
type.

Subsequent commits will update the rest of Wash’s code to reflect the refactor.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Update FUSE to reflect the read refactor

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Update the volume package to reflect the Read refactor

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Update the Docker plugin to reflect the Read refactor

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Update the Kubernetes plugin to reflect the Read refactor

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Update the AWS plugin to reflect the Read refactor

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Update the GCP plugin to reflect the Read refactor

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Update `wash validate` with the Read refactor

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented on plugin/methodWrappers.go:
```diff
@@ -129,14 +129,41 @@ func List(ctx context.Context, p Parent) (*EntryMap, error) {
+		panic("plugin.Read called on a non-readable entry")
+	}
+	data = []byte{}
+	if attr := e.attributes(); attr.HasSize() {
```
It feels weird for a BlockReadable entry to have an unknown size attribute (why would the API let you block-read its content without telling you how big that content is?). Should we panic if we run into that case?

Reason I bring it up here is b/c if we do run into that kind of entry, then the plugin API's responsible for implementing `Read`'s semantics. `Readable` entries w/ the default signature don't need to worry about this because the `entryContentImpl` type takes care of all that.

ekinanp replied:
I'm also thinking about changing `size int64` => `size int`. Reason is b/c `len(data)` is an int.

ekinanp replied:
Is it worth returning `io.EOF` errors? When would it be useful for plugins to provide that information (aside from the block-readable entry w/ unknown size attribute case)? Asking b/c we ignore `io.EOF` in FUSE.

MikaelSmith replied:
I think leaving out `io.EOF` makes sense. I don't think we need it. We would potentially need it if we switched to `OpenDirectIO` mode (as noted in https://godoc.org/bazil.org/fuse/fs#HandleReader), which could help with #609.

MikaelSmith replied:
Oh, I also think we should require a BlockReadable entry to set the size attribute.

ekinanp commented:
I'll also look into https://github.com/puppetlabs/wash/issues/609 in a separate PR.

MikaelSmith commented on plugin/cache.go:
```diff
@@ -208,20 +208,51 @@ func cachedList(ctx context.Context, p Parent) (*EntryMap, error) {
+		case blockReadableSignature:
+			// Go doesn't allow overloaded functions, so external plugin entries will
+			// implement the BlockReadable interface via the blockRead method.
+			var readFunc func(ctx context.Context, size int64, offset int64) ([]byte, error)
```
Maybe make this a type if we're going to use it here and in the `newBlockReadableEntryContent` call.

ekinanp replied:
Like `type readFunc = func(ctx context.Context, size int64, offset int64) ([]byte, error)` ?

MikaelSmith replied:
Yes.

ekinanp replied:
Ok

MikaelSmith commented on plugin/cache_test.go:
```diff
@@ -369,13 +373,92 @@ func (suite *CacheTestSuite) TestCachedListSetEntryID() {
+	expectedRawContent := []byte("some raw content")
+	entry.On("Read", ctx, int64(10), int64(0)).Return(expectedRawContent, nil).Once()
+
+	content, err := cachedRead(ctx, entry)
```
I feel like somewhere in here we should be verifying that the length of content we got back matched the requested length.

MikaelSmith replied:
Or at least actualLength <= requestedLength. Because we're probably going to truncate anything longer than the requested length.

ekinanp replied:
Not sure what you mean. I ended up adding a test-case for `content.size`.

I'm thinking any `actualLength <= requestedLength` checks should be done in `plugin.Read`.

MikaelSmith replied:
Yes, I think you're right.

MikaelSmith replied:
My main observation in the test was that you requested a read of size 10, but get back content longer than that. And nothing fails.

MikaelSmith reviewed (changes requested):
Just a few minor suggestions.

ekinanp commented on plugin/cache.go:
```diff
@@ -208,20 +208,51 @@ func cachedList(ctx context.Context, p Parent) (*EntryMap, error) {
+			}
+			return newEntryContent(rawContent), nil
+		case blockReadableSignature:
+			// Go doesn't allow overloaded functions, so external plugin entries will
```
What do you mean?

MikaelSmith replied:
I mean, is it possible for a type to implement both `Read` signatures for `Readable` and `BlockReadable`? If so, then I don't think I understand your code comment.

ekinanp replied:
Ah ok. Yeah, that is impossible. I'll further clarify the comment.

ekinanp committed:
Ignore io.EOF errors in plugin.Read

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Address PR review feedback for #638

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented:
Updated to ignore `io.EOF` errors, and to also address the current feedback.

ekinanp commented:
Going to add a bit more validation to `plugin.Read` before merging this.

ekinanp committed:
Revert "Ignore io.EOF errors in plugin.Read"

This reverts commit fe59c2bc69cddf2562120290656f5dd8434b9f22. We preserve EOF
because it matches people's expectations of block-readable function behavior.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add some more validation to plugin.Read

plugin.Read now guards against a negative size/offset,
logs a warning if a block-readable entry doesn't set
the size attribute, and also returns an error if the
read data is more than the requested amount.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented:
Updated to restore `io.EOF` and to add some more validation to `plugin.Read`.

ekinanp commented on plugin/methodWrappers.go:
```diff
@@ -153,6 +162,8 @@ func Read(ctx context.Context, e Entry, size int64, offset int64) (data []byte,
 			err = io.EOF
 		}
+	} else if ReadAction().signature(e) == blockReadableSignature {
+		activity.Warnf(ctx, "size attribute not set for block-readable entry %v", e.id())
```
A warning seems fine since this will likely come up when the plugin author's testing their plugin.

ekinanp commented on plugin/methodWrappers.go:
```diff
@@ -163,6 +174,9 @@ func Read(ctx context.Context, e Entry, size int64, offset int64) (data []byte,
 		err = readErr
 	}
+	if actualSize := int64(len(data)); actualSize > size {
+		return nil, fmt.Errorf("requested %v bytes (input was %v), but plugin's API returned %v bytes", size, inputSize, actualSize)
```
This should also come up during plugin author's testing.
